# Pandas 


In [69]:
import pandas as pd 
import numpy as np

In [70]:

# Create a DataFrame manually
data = {'Name': ['Mary', 'Bob', 'Charlie', 'David', 'Eve'],
        'Age': [25, 30, 22, 35, 28],
        'City': ['Nairobi', 'Lagos', 'Juba', 'Accra', 'New York'],
        'Score': [85, 92, 78, 95, 88]}

#Converting a Data to a Dataframe
my_data =pd.DataFrame(data)
print(my_data)

      Name  Age      City  Score
0     Mary   25   Nairobi     85
1      Bob   30     Lagos     92
2  Charlie   22      Juba     78
3    David   35     Accra     95
4      Eve   28  New York     88


In [71]:
my_data.head() #view first 5 elements of the dataframe
# print(my_data.head())
# display(my_data.head())

,Name,Age,City,Score
0,Mary,25,Nairobi,85
1,Bob,30,Lagos,92
2,Charlie,22,Juba,78
3,David,35,Accra,95
4,Eve,28,New York,88


In [72]:
my_data.tail(3) # prints the last 5 elements by default

,Name,Age,City,Score
2,Charlie,22,Juba,78
3,David,35,Accra,95
4,Eve,28,New York,88


In [73]:
# It gives general info abt the dataframe 
# 1. The columns and rows 
# 2. The data types
# 3.The number/quantity  of data -missing data
# 

my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    5 non-null      object
 1   Age     5 non-null      int64 
 2   City    5 non-null      object
 3   Score   5 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 292.0+ bytes


In [74]:
# Describe - The statistical bit about a dataset
# my_data.describe()
print(my_data.describe())

             Age      Score
count   5.000000   5.000000
mean   28.000000  87.600000
std     4.949747   6.580274
min    22.000000  78.000000
25%    25.000000  85.000000
50%    28.000000  88.000000
75%    30.000000  92.000000
max    35.000000  95.000000


In [75]:
# counting the missing data in a dataset and find the sum 
my_data.isnull().sum()

Name     0
Age      0
City     0
Score    0
dtype: int64

In [76]:
# unique values 
print(my_data['City'].unique())

#Count the unique values 
print(my_data['City'].value_counts())


['Nairobi' 'Lagos' 'Juba' 'Accra' 'New York']
City
Nairobi     1
Lagos       1
Juba        1
Accra       1
New York    1
Name: count, dtype: int64


# Using Real DataSet with Pandas

In [78]:
# Loading the Dataset
my_dataset = pd.read_csv('sales_data_sample.csv', encoding='latin-1')
# Display the dataset
display(my_dataset.head())

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,ADDRESSLINE1,ADDRESSLINE2,CITY,STATE,POSTALCODE,COUNTRY,TERRITORY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,2/24/2003 0:00,Shipped,1,2,2003,...,897 Long Airport Avenue,NaN,NYC,NY,10022,USA,NaN,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,5/7/2003 0:00,Shipped,2,5,2003,...,59 rue de l'Abbaye,NaN,Reims,NaN,51100,France,EMEA,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,7/1/2003 0:00,Shipped,3,7,2003,...,27 rue du Colonel Pierre Avia,NaN,Paris,NaN,75508,France,EMEA,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,8/25/2003 0:00,Shipped,3,8,2003,...,78934 Hillside Dr.,NaN,Pasadena,CA,90003,USA,NaN,Young,Julie,Medium
4,10159,49,100.00,14,5205.27,10/10/2003 0:00,Shipped,4,10,2003,...,7734 Strong St.,NaN,San Francisco,CA,NaN,USA,NaN,Brown,Julie,Medium


In [79]:
my_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDERNUMBER       2823 non-null   int64  
 1   QUANTITYORDERED   2823 non-null   int64  
 2   PRICEEACH         2823 non-null   float64
 3   ORDERLINENUMBER   2823 non-null   int64  
 4   SALES             2823 non-null   float64
 5   ORDERDATE         2823 non-null   object 
 6   STATUS            2823 non-null   object 
 7   QTR_ID            2823 non-null   int64  
 8   MONTH_ID          2823 non-null   int64  
 9   YEAR_ID           2823 non-null   int64  
 10  PRODUCTLINE       2823 non-null   object 
 11  MSRP              2823 non-null   int64  
 12  PRODUCTCODE       2823 non-null   object 
 13  CUSTOMERNAME      2823 non-null   object 
 14  PHONE             2823 non-null   object 
 15  ADDRESSLINE1      2823 non-null   object 
 16  ADDRESSLINE2      302 non-null    object 


In [80]:
my_dataset.describe()

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,QTR_ID,MONTH_ID,YEAR_ID,MSRP
count,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.00000,2823.000000
mean,10258.725115,35.092809,83.658544,6.466171,3553.889072,2.717676,7.092455,2003.81509,100.715551
std,92.085478,9.741443,20.174277,4.225841,1841.865106,1.203878,3.656633,0.69967,40.187912
min,10100.000000,6.000000,26.880000,1.000000,482.130000,1.000000,1.000000,2003.00000,33.000000
25%,10180.000000,27.000000,68.860000,3.000000,2203.430000,2.000000,4.000000,2003.00000,68.000000
50%,10262.000000,35.000000,95.700000,6.000000,3184.800000,3.000000,8.000000,2004.00000,99.000000
75%,10333.500000,43.000000,100.000000,9.000000,4508.000000,4.000000,11.000000,2004.00000,124.000000
max,10425.000000,97.000000,100.000000,18.000000,14082.800000,4.000000,12.000000,2005.00000,214.000000


In [81]:
# Data Cleaning
print(my_dataset.isnull().sum())

ORDERNUMBER            0
QUANTITYORDERED        0
PRICEEACH              0
ORDERLINENUMBER        0
SALES                  0
ORDERDATE              0
STATUS                 0
QTR_ID                 0
MONTH_ID               0
YEAR_ID                0
PRODUCTLINE            0
MSRP                   0
PRODUCTCODE            0
CUSTOMERNAME           0
PHONE                  0
ADDRESSLINE1           0
ADDRESSLINE2        2521
CITY                   0
STATE               1486
POSTALCODE            76
COUNTRY                0
TERRITORY           1074
CONTACTLASTNAME        0
CONTACTFIRSTNAME       0
DEALSIZE               0
dtype: int64


In [82]:
# dropna() - Drop rows with missing values
my_dataset = my_dataset.drop(columns=['ADDRESSLINE2', 'STATE', 'POSTALCODE','TERRITORY'])
display(my_dataset)

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,MSRP,PRODUCTCODE,CUSTOMERNAME,PHONE,ADDRESSLINE1,CITY,COUNTRY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,2/24/2003 0:00,Shipped,1,2,2003,...,95,S10_1678,Land of Toys Inc.,2125557818,897 Long Airport Avenue,NYC,USA,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,5/7/2003 0:00,Shipped,2,5,2003,...,95,S10_1678,Reims Collectables,26.47.1555,59 rue de l'Abbaye,Reims,France,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,7/1/2003 0:00,Shipped,3,7,2003,...,95,S10_1678,Lyon Souveniers,+33 1 46 62 7555,27 rue du Colonel Pierre Avia,Paris,France,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,8/25/2003 0:00,Shipped,3,8,2003,...,95,S10_1678,Toys4GrownUps.com,6265557265,78934 Hillside Dr.,Pasadena,USA,Young,Julie,Medium
4,10159,49,100.00,14,5205.27,10/10/2003 0:00,Shipped,4,10,2003,...,95,S10_1678,Corporate Gift Ideas Co.,6505551386,7734 Strong St.,San Francisco,USA,Brown,Julie,Medium
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2818,10350,20,100.00,15,2244.40,12/2/2004 0:00,Shipped,4,12,2004,...,54,S72_3212,Euro Shopping Channel,(91) 555 94 44,"C/ Moralzarzal, 86",Madrid,Spain,Freyre,Diego,Small
2819,10373,29,100.00,1,3978.51,1/31/2005 0:00,Shipped,1,1,2005,...,54,S72_3212,"Oulu Toy Supplies, Inc.",981-443655,Torikatu 38,Oulu,Finland,Koskitalo,Pirkko,Medium
2820,10386,43,100.00,4,5417.57,3/1/2005 0:00,Resolved,1,3,2005,...,54,S72_3212,Euro Shopping Channel,(91) 555 94 44,"C/ Moralzarzal, 86",Madrid,Spain,Freyre,Diego,Medium
2821,10397,34,62.24,1,2116.16,3/28/2005 0:00,Shipped,1,3,2005,...,54,S72_3212,Alpha Cognac,61.77.6555,1 rue Alsace-Lorraine,Toulouse,France,Roulet,Annette,Small


#  **Calculations**

In [84]:
# SALES BY COUNTRIES
print(my_dataset.groupby ('COUNTRY')['SALES'].sum().sort_values(ascending=False))

COUNTRY
USA            3627982.83
Spain          1215686.92
France         1110916.52
Australia       630623.10
UK              478880.46
Italy           374674.31
Finland         329581.91
Norway          307463.70
Singapore       288488.41
Denmark         245637.15
Canada          224078.56
Germany         220472.09
Sweden          210014.21
Austria         202062.53
Japan           188167.81
Switzerland     117713.56
Belgium         108412.62
Philippines      94015.73
Ireland          57756.43
Name: SALES, dtype: float64


### A summary of SALES BY COUNTRY 
* There are **19 countries**
* The **USA** has the highest number of SALES by **3627982.83**
* The **Ireland** records the lowest number of SALES by **57756.43**

In [86]:
# Group Dealsize by mean 
average_sale_by_dealsize = my_dataset.groupby("DEALSIZE")["SALES"].mean()
average_sale_by_dealsize

DEALSIZE
Large     8293.753248
Medium    4398.433699
Small     2061.682800
Name: SALES, dtype: float64

### A summary of DealSize Analysis 
* The large DEALSIZE has the highest average SALES with **8293.75**
*  The medium DEALSIZE has a moderate average number of SALES with **4398.43**
*  The small DEALSIZE has the lowest average number of SALES with **2061.68**


# **Numpy**

In [89]:
# Import libraries 
import pandas as pd
import numpy as np 

In [90]:
# Create a NumPy array
data = np.array([['Alice', 25, 'Nairobi', 85],
                 ['Bob', 30, 'Lagos', 92],
                 ['Charlie', 22, 'Juba', 78],
                 ['David', 35, 'Accra', 95],
                 ['Eve', 28, 'New York', 88]])

# Create column names
columns = ['Name', 'Age', 'City', 'Score']

# Create a Pandas DataFrame from the NumPy array
my_df = pd.DataFrame(data=data, columns=columns)

In [91]:
my_df.head()

,Name,Age,City,Score
0,Alice,25,Nairobi,85
1,Bob,30,Lagos,92
2,Charlie,22,Juba,78
3,David,35,Accra,95
4,Eve,28,New York,88


# **Data Analysis with Numpy**

In [93]:
# Converting values of Ages to integers/numbers 
# Converting values of Scores to integers/numbers 

ages = my_df['Age'].to_numpy(dtype=int)
scores = my_df['Score'].to_numpy(dtype=int)


In [94]:
# Calculating the average score 
average_score = np.mean(scores)
average_score

87.6

In [95]:
#Calculating the Max Score 
max_score = np.max(scores)
max_score

95

In [96]:
#Calculating the Min Score 
min_score = np.min(scores)
min_score

78

In [97]:
# Calculating the Median Score
median_score = np.median(scores)
median_score

88.0

In [98]:
# Calculating the Standard Deviation
std_score = np.std(scores)
print(f"Standard Deviation:{std_score:.2f}")

Standard Deviation:5.89


In [99]:
# Calculating Total Scores
sum_scores = np.sum(scores)
sum_scores

438

### **A Summary of Numpy Analysis**
* The **average** Scores is **87.6**
* The **Max** Score is **95**
* The **Min** Score is **78**
* The **Median** Score is **88**
* The **Standard Deviation** Score is **5.89**
* The **Sum total** of Score is **438**

# **Data Analysis using a Real Dataset**

In [128]:
# Loading the Dataset
my_dataset = pd.read_csv('sales_data_sample.csv', encoding='latin-1')
# Display the dataset
display(my_dataset.head())

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,ADDRESSLINE1,ADDRESSLINE2,CITY,STATE,POSTALCODE,COUNTRY,TERRITORY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,2/24/2003 0:00,Shipped,1,2,2003,...,897 Long Airport Avenue,NaN,NYC,NY,10022,USA,NaN,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,5/7/2003 0:00,Shipped,2,5,2003,...,59 rue de l'Abbaye,NaN,Reims,NaN,51100,France,EMEA,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,7/1/2003 0:00,Shipped,3,7,2003,...,27 rue du Colonel Pierre Avia,NaN,Paris,NaN,75508,France,EMEA,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,8/25/2003 0:00,Shipped,3,8,2003,...,78934 Hillside Dr.,NaN,Pasadena,CA,90003,USA,NaN,Young,Julie,Medium
4,10159,49,100.00,14,5205.27,10/10/2003 0:00,Shipped,4,10,2003,...,7734 Strong St.,NaN,San Francisco,CA,NaN,USA,NaN,Brown,Julie,Medium


In [130]:
# Converting SALES datatype to Integer 
sales = my_dataset['SALES'].to_numpy()
# Convert QUANTITYORDERED datatype to Integer
quantity_ordered = my_dataset['QUANTITYORDERED'].to_numpy()

In [132]:
# Calculating Total SALES
total_sales = np.sum(sales)
total_sales

10032628.85

In [134]:
# Calculating the Max Sales
max_sales = np.max(sales)
max_sales

14082.8

In [136]:
# Calculating the Min Sales
min_sales = np.min(sales)
min_sales

482.13

In [142]:
# Calculating the Average Sales
average_sales = np.mean(sales)
print(f"{average_sales:.2f}")

3553.89


### Summary of Numpy Analysis using Real Dataset 
* The **Total** Sales is **10032628.85**
* The **Max** Sales is  **14082.8**
* The **Min**  Sales is  **482.3**
* The **Average** Sales is **3553.89**